openstack_hypervisor/hooks.py

```diff
@@ -1690,6 +1716,157 @@ def _configure_libvirt_tls(snap: Snap) -> None:
+    nova_key = nova_private_dir / "serverkey.pem"
+    _template_tls_file(cert, nova_cert, [])
+    _template_tls_file(key, nova_key, [], PRIVATE_PERMS)
+
+    # Nova fileserver CA (for client verification)
+    nova_ca = nova_dir / "ca-cert.pem"
+    _template_tls_file(cacert, nova_ca, [])
+
+    common_base = snap.paths.common
+    tls_parent = common_base / Path("apache-webdav")
+    webdav_tls_dir = tls_parent / "tls"
+    for path in (tls_parent, webdav_tls_dir):
+        path.mkdir(parents=True, exist_ok=True)
+        path.chmod(0o755)
+
+    apache_cert = webdav_tls_dir / "servercert.pem"
+    apache_key = webdav_tls_dir / "serverkey.pem"
+    apache_ca = webdav_tls_dir / "ca-cert.pem"
+
+    shutil.copyfile(nova_cert, apache_cert)
+    shutil.copyfile(nova_key, apache_key)
+    shutil.copyfile(nova_ca, apache_ca)
+
+    apache_cert.chmod(0o644)
+    apache_key.chmod(0o644)
+    apache_ca.chmod(0o644)
+
+    _configure_webdav_apache(snap)
+
+
+def _configure_webdav_apache(snap: Snap) -> None:
+    """Configure Apache-based WebDAV endpoint with mTLS."""
+    apache_cfg_dir = snap.paths.common / Path("etc/apache2")
+    apache_cfg_dir.mkdir(parents=True, exist_ok=True)
+    for ancestor in (apache_cfg_dir, apache_cfg_dir.parent):
+        if ancestor.exists():
+            ancestor.chmod(0o755)
+    webdav_cfg_path = apache_cfg_dir / "webdav.conf"
+    mime_types_path = apache_cfg_dir / "mime.types"
+    server_root = snap.paths.snap / Path("usr/lib/apache2")
+    modules_dir = server_root / "modules"
+    if not mime_types_path.exists():
+        src_candidates = [
+            Path("/etc/mime.types"),
+            snap.paths.snap / Path("etc/mime.types"),
+            snap.paths.snap / Path("usr/share/misc/mime.types"),
+        ]
+        source = next((p for p in src_candidates if p.exists()), None)
+        if source:
+            shutil.copy(source, mime_types_path)
+        else:
+            mime_types_path.write_text(
+                "text/plain\t txt\napplication/octet-stream\tbin\n", encoding="utf-8"
+            )
+
+    webdav_root_fs = LAYOUT_BASE / Path("var/webdav")
+    dav_lock_dir_fs = LAYOUT_BASE / Path("var/webdav-lock")
+    log_dir_fs = LAYOUT_BASE / Path("log/apache2")
+    run_dir_fs = LAYOUT_BASE / Path("run/apache2")
+    mime_types_path_fs = LAYOUT_BASE / Path("etc/apache2/mime.types")
+
+    def module(name: str) -> Path:
+        return modules_dir / f"mod_{name}.so"
+
+    dav_lock_db = dav_lock_dir_fs / "DavLock"
+    error_log = log_dir_fs / "error.log"
+    access_log = log_dir_fs / "access.log"
+    client_log = log_dir_fs / "ssl_client.log"
+    pid_file = run_dir_fs / "apache2.pid"
+
+    # The WebDAV endpoint listens on all interfaces so peer compute nodes can
+    # reach it during migrations. Restricting to loopback would break inter-node
+    # disk transfer flows. Timeout is intentionally long (3h) to cover large
+    # disk uploads over constrained links.
+    config = (
```

Comment:
Is it possible to templatize the conf file and put it in templates and reuse the existing mechanism?

Reply:
This should still be a template and not a huge blob written into the hooks file.

All other templated files are constructed on the fly based on parameters inserted via the Jinja Context (i.e. python variables).